andela/ah-backend-invictus
# #165273480 Add user like and or dislike article

**What does this PR do?**
It enables a user who has logged in to be able to like and or dislike their own or other people's articles

*Tasks*
- Edit model for Articles to add two fields likes_count and dislikes_count
- Add model for Like_Dislike which takes in foreign keys of User model and Articles and has a like_or_dislike field
- Add tests for like and dislike article functionality
- Add a serializer for the Like_Dislike model
- Edit article views to add two view endpoint
- Edit article urls.py file to add the two URLs for the like and dislike endpoint 
- Add validation for the user not liking or disliking the same article twice. If the endpoint for either like or dislike is hit again, the like or dislike is revoked and the likes count or dislike count in the article is updated.
- Add functionality for if an authenticated user likes a previously disliked article, the dislike is deleted, dislike count reduces by one on the article and the like count increases by one with the like being saved and vice versa for if the user dislikes a previously liked article.



*How to test out this PR?*
Log in a user and go to access the following endpoints while passing the token in the header for existing articles to either like or dislike them.

*Endpoints*
  ```/api/article/<article_id>/like```       Likes a specific article by its id
 ```/api/article/<article_id>/dislike```   Dislikes specific article
```/api/article/<article_id>```                To check the article @ fields likes_count, dislikes_count

*Like an article*
There is no data in the request body. 
The following is the expected response body.
```
{
    "success": "You have successfully liked this article."
}
```
The expected response if you try to like the same article twice
```
{
    "message": "Your like has been revoked"
}
```
If the article is not found, this is the expected response body,
```
{
      "error": "That article is not found"
}
``` 

For disliking, the expected responses are similar. Simply expect dislike in the space of like
This then reduces the count in the article attribute likes_count

*View article like count after like*
The image below is from when you search article by id. View likes_count field.
 ```/api/article/<article_id>/``` To view article
<img width="443" alt="Screen Shot 2019-05-01 at 7 00 47 PM" src="https://user-images.githubusercontent.com/44635784/57026865-7c910c00-6c43-11e9-8e9f-d9f52f4547b1.png">

If the user hits like endpoint then immediately hits dislike endpoint, the expected response is

```
{
    "success": "Your like for the article has changed to a dislike."
}
```
 

[Starts #165273480]
